### ISSUE TITLE
[kube-state-metrics] kubernetes_state.job.succeeded/failed wildly out of wack

### ISSUE THREAD

**Issue by SleepyBrett** [DataDog/datadog-agent#4572]
**Output of the info page (if this is a bug)**
```
==============
Agent (v6.15.0)
===============

  Status date: 2019-12-06 16:34:47.824750 UTC
  Agent start: 2019-11-15 03:36:36.073832 UTC
  Pid: 341
  Go Version: go1.12.9
  Python Version: 2.7.16
  Check Runners: 4
  Log Level: WARNING

  Paths
  =====
    Config File: /etc/datadog-agent/datadog.yaml
    conf.d: /etc/datadog-agent/conf.d
    checks.d: /etc/datadog-agent/checks.d

  Clocks
  ======
    System UTC time: 2019-12-06 16:34:47.824750 UTC

  Host Info
  =========
    bootTime: 2019-11-15 03:34:15.000000 UTC
    kernelVersion: 4.19.68-coreos
    os: linux
    platform: debian
    platformFamily: debian
    platformVersion: 10.1
    procs: 71
    uptime: 2m23s
    virtualizationRole: guest
    virtualizationSystem: xen

  Hostnames
  =========
    ec2-hostname: ip-172-16-207-57.us-west-2.compute.internal
    host_aliases: [ip-172-16-207-57.us-west-2.compute.internal]
    hostname: i-00fc1de3f7cbadca3
    instance-id: i-00fc1de3f7cbadca3
    socket-fqdn: dd-prod-datadog-szzcd
    socket-hostname: dd-prod-datadog-szzcd
    host tags:
      [kube_cloud_provider:aws
      kube_cluster_id:steel
      kube_env:prod
      kube_region:us-west-2]
    hostname provider: aws
    unused hostname providers:
      configuration/environment: hostname is empty
      gce: unable to retrieve hostname from GCE: status code 404 trying to GET http://169.254.169.254/computeMetadata/v1/instance/hostname

=========
Collector
=========

  Running Checks
  ==============

    coredns (1.2.0)
    ---------------
      Instance ID: coredns:eab009ef2bf6f24c [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/coredns.d/auto_conf.yaml
      Total Runs: 124,073
      Metric Samples: Last Run: 488, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 124,073
      Average Execution Time : 44ms


    cpu
    ---
      Instance ID: cpu [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/cpu.d/conf.yaml.default
      Total Runs: 124,073
      Metric Samples: Last Run: 6, Total: 744,432
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s


    cri
    ---
      Instance ID: cri [ERROR]
      Configuration Source: file:/etc/datadog-agent/conf.d/cri.d/conf.yaml.default
      Total Runs: 124,072
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s
      Error: permanent failure in criutil: retry number exceeded
      No traceback
      Warning: Error initialising check: permanent failure in criutil: retry number exceeded


    disk (2.5.1)
    ------------
      Instance ID: disk:e5dffb8bef24336f [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/disk.d/conf.yaml.default
      Total Runs: 124,073
      Metric Samples: Last Run: 208, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 117ms


    file_handle
    -----------
      Instance ID: file_handle [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/file_handle.d/conf.yaml.default
      Total Runs: 124,072
      Metric Samples: Last Run: 5, Total: 620,360
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s


    io
    --
      Instance ID: io [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/io.d/conf.yaml.default
      Total Runs: 124,073
      Metric Samples: Last Run: 156, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 1ms


    kubelet (3.3.4)
    ---------------
      Instance ID: kubelet:d884b5186b651429 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/kubelet.d/conf.yaml.default
      Total Runs: 124,072
      Metric Samples: Last Run: 2,363, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 4, Total: 496,288
      Average Execution Time : 1.091s


    kubernetes_state (4.7.1)
    ------------------------
      Instance ID: kubernetes_state:853b0bed6c1a5803 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/kubernetes_state.d/auto_conf.yaml
      Total Runs: 124,031
      Metric Samples: Last Run: 31,756, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1,406, Total: 1 M
      Average Execution Time : 2.677s


    load
    ----
      Instance ID: load [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/load.d/conf.yaml.default
      Total Runs: 124,073
      Metric Samples: Last Run: 6, Total: 744,438
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s


    memory
    ------
      Instance ID: memory [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/memory.d/conf.yaml.default
      Total Runs: 124,072
      Metric Samples: Last Run: 17, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s


    network (1.11.5)
    ----------------
      Instance ID: network:e0204ad63d43c949 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/network.d/conf.yaml.default
      Total Runs: 124,073
      Metric Samples: Last Run: 505, Total: 1 M
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 11ms


    ntp
    ---
      Instance ID: ntp:d884b5186b651429 [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/ntp.d/conf.yaml.default
      Total Runs: 2,068
      Metric Samples: Last Run: 0, Total: 0
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 1, Total: 2,068
      Average Execution Time : 20.013s


    uptime
    ------
      Instance ID: uptime [OK]
      Configuration Source: file:/etc/datadog-agent/conf.d/uptime.d/conf.yaml.default
      Total Runs: 124,072
      Metric Samples: Last Run: 1, Total: 124,072
      Events: Last Run: 0, Total: 0
      Service Checks: Last Run: 0, Total: 0
      Average Execution Time : 0s

========
JMXFetch
========

  Initialized checks
  ==================
    no checks

  Failed checks
  =============
    no checks

=========
Forwarder
=========

  Transactions
  ============
    CheckRunsV1: 124,072
    Dropped: 0
    DroppedOnInput: 0
    Events: 0
    HostMetadata: 0
    IntakeV1: 13,441
    Metadata: 0
    Requeued: 5
    Retried: 5
    RetryQueueSize: 0
    Series: 0
    ServiceChecks: 0
    SketchSeries: 0
    Success: 706,682
    TimeseriesV1: 124,072

  Transaction Errors
  ==================
    Total number: 1
    Errors By Type:

  API Keys status
  ===============
    API key ending with 2c443: API Key valid

==========
Endpoints
==========
  https://app.datadoghq.com - API Key ending with:
      - 2c443

==========
Logs Agent
==========

  Logs Agent is not running

=========
Aggregator
=========
  Checks Metric Sample: 4 G
  Dogstatsd Metric Sample: 310,182
  Event: 1
  Events Flushed: 1
  Number Of Flushes: 124,072
  Series Flushed: 4 G
  Service Check: 173.1 M
  Service Checks Flushed: 173.2 M

=========
DogStatsD
=========
  Event Packets: 0
  Event Parse Errors: 0
  Metric Packets: 310,181
  Metric Parse Errors: 0
  Service Check Packets: 0
  Service Check Parse Errors: 0
  Udp Bytes: 17.2 M
  Udp Packet Reading Errors: 0
  Udp Packets: 310,182
  Uds Bytes: 0
  Uds Origin Detection Errors: 0
  Uds Packet Reading Errors: 0
  Uds Packets: 0

```

It should be noted that this particular agent is scraping only part of the metrics, because of dd agent issues we've divided our KSM into two pods one for all pod metrics and one for all other metrics. They are located on different nodes and therefore scraped by different agents.

**Describe what happened:**

Kubernetes 1.16.2
KSM 1.18.0
DD-agent 6.15.0

failed/succeeded job counts do not line up *at all*.

![image](https://user-images.githubusercontent.com/279521/70339934-9e1f2b00-1804-11ea-94ca-fc7bafa2f4a4.png)
![image](https://user-images.githubusercontent.com/279521/70339944-a37c7580-1804-11ea-9293-f8bfc3162157.png)

I'm not sure what explains any metrics at all flowing, i would expect a fairly binary outcome here.

**Describe what you expected:**

I expect my stats to line up.

It was mentioned by @CharlyF that potentially the problem here is that you haven't run any tests against KSM 1.18.0. This is quite possibly the problem I bisecting through KSM to see if I can't quick patch this.

Here are my expectations, I would like you guys to do a better job publishing tested versions of major components for each integration. I'm growing frustrated with the firedrill of datadog issues whenever we perform an upgrade.

Just a few weeks back I found this bug https://github.com/DataDog/integrations-core/issues/4923.

The post-mortem timeline for that is:
  * 02/26/19 Deprication of pod/container_name in the kubelet is announced (as far as I can tell) in the 1.14.0-beta.1 patch notes. At this point both the new pod/container labels as well as the old pod_name/container_name labels are on the metrics produced from the kubelet.
  * 09/07/19 Kube 1.16.0-Alpha.3 is released in the patch notes it's made clear that the labels are now fully depricated.
  * 09/18/19 Kube 1.16.0 is released
  * 10/29/19 I create the issue after discovering that the kubelet integration was still looking for the old metric labels.
  * 10/30/19 To your credit your team acted quickly to get a fix in place
  * 11/05/19 That fix makes it to an agent (6.15.0)

Kubernetes is moving fast but you had 9 months to react.

**Steps to reproduce the issue:**

Run the above noted configuration.

**Additional environment details (Operating System, Cloud provider, etc):**
NA

**Comment by CharlyF** [comment]
Thanks for opening this issue @SleepyBrett. 
The best way forward is to verify that the values you expect are properly listed in the KSM output, our source of truth. As you mentioned:
> It should be noted that this particular agent is scraping only part of the metrics, because of dd agent issues we've divided our KSM into two pods one for all pod metrics and one for all other metrics. They are located on different nodes and therefore scraped by different agents.

Could you share more details on the issues you had with the Datadog Agent ? Are we talking large memory footprint or something else ?
With regards to how you separated the KSM metrics, do you mean splitting the collectors or are you also using the autosharding feature ?

Could you exec in the Agent collecting the jobs metrics (I am assuming you just split the collectors) and curl the ksm endpoint ? I am specially interested in the `kube_job_failed` and `kube_job_status_failed` metrics ([two different metrics in our check](https://github.com/DataDog/integrations-core/blob/master/kubernetes_state/datadog_checks/kubernetes_state/kubernetes_state.py#L105-L106), but I am not familiar with how Prometheus evaluates both and from your screenshot it seems like only the status_failed is plotted).

Best,
.C

**Comment by SleepyBrett** [comment]
Large memory footprints and timeouts related to trying to process a whole KSM, the sharding was actually potentially your idea (at least your team's idea) (https://github.com/DataDog/datadog-agent/issues/1665, https://github.com/DataDog/datadog-agent/issues/1853)... though we were also working with a sales support engineer on this issue.

We had no access to the autosharding feature at the time so we sharded by collector (pods in one pod, everything else in the other). I did notice the addition to a beta feature for sharding and have it on a list to try out.

```
...
kube_job_failed{namespace="cvp",job_name="nonprod-automl-deployer-1575657000",condition="true"} 1
kube_job_failed{namespace="cvp",job_name="nonprod-automl-deployer-1575657000",condition="false"} 0
kube_job_failed{namespace="cvp",job_name="nonprod-automl-deployer-1575657000",condition="unknown"} 0
...
kube_job_status_failed{namespace="devportal",job_name="scraper-1575547200"} 0
kube_job_status_failed{namespace="dsdatafnd",job_name="no-meta-gp-zb4tdtp-delete-job"} 1
...
```
enough for you?

**Comment by DylanLovesCoffee** [comment]
@SleepyBrett Just giving a heads up that the fix for cronjobs was released with v6.16.x of the agent via https://github.com/DataDog/integrations-core/pull/4943, so if upgrading is a possibility that should provide a first step.

**Comment by SleepyBrett** [comment]
6.16.1 seems to have resolved this. We've also changed our sharding methodology to the ksm built in option (5 pods right now).
